### PR TITLE
test: add, refactor ipv6 network, vpc tests

### PR DESCRIPTION
**Pull request by shwstppr**
### Description
- Refactor IPv6 related tests
- Adds smoke test for IPv4 network to IPv6 upgrade
- Adds smoke test for IPv6 VPC

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [x] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by shwstppr** [comment]
@rohityadavcloud acs-robot calling UI build when there are no UI changes in PR ^^

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

**Review by rohityadavcloud — APPROVED** [review]

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Review by DaanHoogland — APPROVED** [review]

**Comment by shwstppr** [comment]
@blueorangutan test

@rohityadavcloud @weizhouapache @DaanHoogland added smoke tests may add ~1.25hrs to smoke test run. Does it make sense to move some of the test cases to a test file in component  dir?

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
I think it makes sense to only leave miniml tests in the smoke dir, @shwstppr . The rest we´d only want to run on subject specific PRs and releases.
€0.02

**Comment by weizhouapache** [comment]
> I think it makes sense to only leave miniml tests in the smoke dir, @shwstppr . The rest we´d only want to run on subject specific PRs and releases. €0.02

agree with @DaanHoogland 
we should run the component tests at least once a month (we need to fix all component tests before run, or move failed component tests to a separated folder)

**Comment by shwstppr** [comment]
Refactored tests. Now only basic test for IPv6 isolated nw and VPC is there in /smoke. /component has all the added Marvin tests.

@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@blueorangutan test

File: server/src/test/java/com/cloud/network/Ipv6ServiceImplTest.java [hunk 1]
```diff
@@ -0,0 +1,781 @@
...
+        prefixes.add(prefix);
+        prefixes.add(prefix);
+        Mockito.when(dataCenterGuestIpv6PrefixDao.listByDataCenterId(zoneId)).thenReturn(prefixes);
+        Pair<Integer, Integer> results = ipv6Service.getUsedTotalIpv6SubnetForZone(zoneId);
+        Assert.assertEquals(4, results.first().intValue());
+        Assert.assertEquals(8, results.second().intValue());
+    }
+
+    @Test(expected = ResourceAllocationException.class)
+    @DB
```
Comment (DaanHoogland) [inline server/src/test/java/com/cloud/network/Ipv6ServiceImplTest.java:219]: Unit tests using DB are a bit worrying. Do these tests reallly need this annotation?
and if so; Are these tests really needed/not covered by the integration tests?

File: test/integration/smoke/test_network_ipv6.py [hunk 2]
```diff
@@ -665,14 +209,16 @@ def setUp(self):
...
         self.apiclient = self.testClient.getApiClient()
         self.dbclient = self.testClient.getDbConnection()
+        self.thread = None
         self.cleanup = []
         return
 
     def tearDown(self):
         try:
+            if self.thread and self.thread.is_alive():
+                self.thread.join(5*60)
```
Comment (DaanHoogland) [inline test/integration/smoke/test_network_ipv6.py:219]: with this extra code we run the test of the cleanup not being reached. Please see the construct I sugested in other tests to apply it here as well.

File: test/integration/smoke/test_vpc_ipv6.py [hunk 3]
```diff
@@ -0,0 +1,911 @@
...
+    def tearDownClass(cls):
+        if cls.initial_ipv6_offering_enabled != None:
+            Configurations.update(cls.apiclient,
+                ipv6_offering_config_name,
+                cls.initial_ipv6_offering_enabled)
+        super(TestIpv6Vpc, cls).tearDownClass()
+        if cls.test_ipv6_guestprefix != None:
+            cmd = deleteGuestNetworkIpv6Prefix.deleteGuestNetworkIpv6PrefixCmd()
+            cmd.id = cls.test_ipv6_guestprefix.id
+            cls.apiclient.deleteGuestNetworkIpv6Prefix(cmd)
```
Comment (DaanHoogland) [inline test/integration/smoke/test_vpc_ipv6.py:179]:
```suggestion
    def tearDownClass(cls):
        try:
            if cls.initial_ipv6_offering_enabled != None:
                Configurations.update(cls.apiclient,
                     ipv6_offering_config_name,
                     cls.initial_ipv6_offering_enabled)
        finally:
            super(TestIpv6Vpc, cls).tearDownClass()
        if cls.test_ipv6_guestprefix != None:
            cmd = deleteGuestNetworkIpv6Prefix.deleteGuestNetworkIpv6PrefixCmd()
            cmd.id = cls.test_ipv6_guestprefix.id
            cls.apiclient.deleteGuestNetworkIpv6Prefix(cmd)
```

should GuestNetworkIpv6Prefix be added as class in base.py?

File: test/integration/smoke/test_vpc_ipv6.py [hunk 4]
```diff
@@ -0,0 +1,911 @@
...
+    def tearDown(self):
+        try:
+            if self.thread and self.thread.is_alive():
+                self.thread.join(5*60)
+            #Clean up, terminate the created templates
+            # cleanup_resources(self.apiclient, reversed(self.cleanup))
+
+        except Exception as e:
+            raise Exception("Warning: Exception during cleanup : %s" % e)
+        return
```
Comment (DaanHoogland) [inline test/integration/smoke/test_vpc_ipv6.py:242]:
```suggestion
    def tearDown(self):
        try:
            if self.thread and self.thread.is_alive():
                self.thread.join(5*60)
        except Exception as e:
            raise Exception("Warning: Exception during cleanup : %s" % e)
        finally:
            super(TestIpv6Vpc, self).tearDown()
        return
```

File: test/integration/smoke/test_vpc_ipv6.py [hunk 5]
```diff
@@ -0,0 +1,911 @@
...
+                "SourceNat": {
+                    "RedundantRouter": 'true'
+                },
+            }
+        if is_ipv6:
+            off_service["internetprotocol"] = "dualstack"
+        vpc_offering = VpcOffering.create(
+            self.apiclient,
+            off_service
+        )
```
Comment (DaanHoogland) [inline test/integration/smoke/test_vpc_ipv6.py:270]:
```suggestion
        vpc_offering = VpcOffering.create(
            self.apiclient,
            off_service
        )
        self.cleanup.append(vpc_offering)
```

File: test/integration/smoke/test_vpc_ipv6.py [hunk 6]
```diff
@@ -0,0 +1,911 @@
...
+        vpc_offering.update(self.apiclient, state='Enabled')
+        return vpc_offering
+
+    def createIpv4VpcOffering(self, is_redundant=False):
+        self.vpc_offering = self.createVpcOfferingInternal(is_redundant, False)
+        self.cleanup.append(self.vpc_offering)
```
Comment (DaanHoogland) [inline test/integration/smoke/test_vpc_ipv6.py:276]:
```suggestion
```

File: test/integration/smoke/test_vpc_ipv6.py [hunk 7]
```diff
@@ -0,0 +1,911 @@
...
+
+    def createIpv6VpcOffering(self, is_redundant=False):
+        self.vpc_offering = self.createVpcOfferingInternal(is_redundant, True)
+        self.cleanup.append(self.vpc_offering)
```
Comment (DaanHoogland) [inline test/integration/smoke/test_vpc_ipv6.py:280]:
```suggestion
```

File: test/integration/smoke/test_vpc_ipv6.py [hunk 8]
```diff
@@ -0,0 +1,911 @@
...
+
+    def createIpv6VpcOfferingForUpdate(self, is_redundant=False):
+        self.vpc_offering_update = self.createVpcOfferingInternal(is_redundant, True)
+        self.cleanup.append(self.vpc_offering_update)
```
Comment (DaanHoogland) [inline test/integration/smoke/test_vpc_ipv6.py:284]:
```suggestion
```

File: test/integration/smoke/test_vpc_ipv6.py [hunk 9]
```diff
@@ -0,0 +1,911 @@
...
+                supportedServices = off_service["supportedservices"].split(",")
+                supportedServices.remove("Lb")
+                off_service["supportedservices"] = ",".join(supportedServices)
+        if is_ipv6:
+            off_service["internetprotocol"] = "dualstack"
+        network_offering = NetworkOffering.create(
+            self.apiclient,
+            off_service,
+            conservemode=False
+        )
```
Comment (DaanHoogland) [inline test/integration/smoke/test_vpc_ipv6.py:303]:
```suggestion
        network_offering = NetworkOffering.create(
            self.apiclient,
            off_service,
            conservemode=False
        )
        self.cleanup.append(network_offering)
```

File: test/integration/smoke/test_vpc_ipv6.py [hunk 10]
```diff
@@ -0,0 +1,911 @@
...
+        network_offering.update(self.apiclient, state='Enabled')
+        return network_offering
+
+    def createIpv4NetworkTierOffering(self):
+        self.network_offering = self.createNetworkTierOfferingInternal(False)
+        self.cleanup.append(self.network_offering)
```
Comment (DaanHoogland) [inline test/integration/smoke/test_vpc_ipv6.py:309]:
```suggestion
```

File: test/integration/smoke/test_vpc_ipv6.py [hunk 11]
```diff
@@ -0,0 +1,911 @@
...
+
+    def createIpv6NetworkTierOffering(self, remove_lb=True):
+        self.network_offering = self.createNetworkTierOfferingInternal(True)
+        self.cleanup.append(self.network_offering)
```
Comment (DaanHoogland) [inline test/integration/smoke/test_vpc_ipv6.py:313]:
```suggestion
```

File: test/integration/smoke/test_vpc_ipv6.py [hunk 12]
```diff
@@ -0,0 +1,911 @@
...
+
+    def createIpv6NetworkTierOfferingForUpdate(self):
+        self.network_offering_update = self.createNetworkTierOfferingInternal(True)
+        self.cleanup.append(self.network_offering_update)
```
Comment (DaanHoogland) [inline test/integration/smoke/test_vpc_ipv6.py:317]:
```suggestion
```

File: test/integration/smoke/test_vpc_ipv6.py [hunk 13]
```diff
@@ -0,0 +1,911 @@
...
+        service = self.services["vpc"]
+        service["cidr"] = cidr
+        vpc = VPC.create(
+            self.apiclient,
+            self.services["vpc"],
+            vpcofferingid=self.vpc_offering.id,
+            zoneid=self.zone.id,
+            account=self.account.name,
+            domainid=self.account.domainid
+        )
```
Comment (DaanHoogland) [inline test/integration/smoke/test_vpc_ipv6.py:329]:
```suggestion
        vpc = VPC.create(
            self.apiclient,
            self.services["vpc"],
            vpcofferingid=self.vpc_offering.id,
            zoneid=self.zone.id,
            account=self.account.name,
            domainid=self.account.domainid
        )
        self.cleanup.append(vpc)
```

File: test/integration/smoke/test_vpc_ipv6.py [hunk 14]
```diff
@@ -0,0 +1,911 @@
...
+        NetworkACL.create(self.apiclient,
+            services=rule,
+            aclid=acl.id
+        )
+        self.vpcAllowAllAclDetailsMap[vpc.id] = acl.id
+        return vpc
+
+    def deployVpc(self):
+        self.vpc = self.deployAllowAllVpcInternal(VPC_DATA["cidr"])
+        self.cleanup.append(self.vpc)
```
Comment (DaanHoogland) [inline test/integration/smoke/test_vpc_ipv6.py:355]:
```suggestion
```

File: test/integration/smoke/test_vpc_ipv6.py [hunk 15]
```diff
@@ -0,0 +1,911 @@
...
+            service,
+            self.account.name,
+            self.account.domainid,
+            networkofferingid=network_offering_id,
+            vpcid=vpc_id,
+            zoneid=self.zone.id,
+            gateway=tier_gateway,
+            netmask=tier_netmask,
+            aclid=acl_id
+        )
```
Comment (DaanHoogland) [inline test/integration/smoke/test_vpc_ipv6.py:375]:
```suggestion
        network = Network.create(
            self.apiclient,
            service,
            self.account.name,
            self.account.domainid,
            networkofferingid=network_offering_id,
            vpcid=vpc_id,
            zoneid=self.zone.id,
            gateway=tier_gateway,
            netmask=tier_netmask,
            aclid=acl_id
        )
        self.cleanup.append(network)
```

File: test/integration/smoke/test_vpc_ipv6.py [hunk 16]
```diff
@@ -0,0 +1,911 @@
...
+        return network
+
+    def deployNetworkTier(self):
+        self.network = self.deployNetworkTierInternal(
+            self.network_offering.id,
+            self.vpc.id,
+            VPC_DATA["tier1_gateway"],
+            VPC_DATA["tier_netmask"]
+        )
+        self.cleanup.append(self.network)
```
Comment (DaanHoogland) [inline test/integration/smoke/test_vpc_ipv6.py:385]:
```suggestion
```

**Review by DaanHoogland — COMMENTED** [review]
more then 50 comments, but only 3 or 4 conceptually

**Comment by rohityadavcloud** [comment]
cc @shwstppr pl check that the base branch changed to 4.17 causes some extra changes/commits (rebase/merge 4.17 branch on your PR branch). I think since you're adding tests, the 4.17 branch can benefit from this which has the ipv6 feature.

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@blueorangutan package

File: test/integration/smoke/test_vpc_ipv6.py [hunk 17]
```diff
@@ -0,0 +1,901 @@
...
+                off_service["supportedservices"] = ",".join(supportedServices)
+        if is_ipv6:
+            off_service["internetprotocol"] = "dualstack"
+        network_offering = NetworkOffering.create(
+            self.apiclient,
+            off_service,
+            conservemode=False
+        )
+        network_offering.update(self.apiclient, state='Enabled')
+        self.cleanup.append(network_offering)
```
Comment (DaanHoogland) [inline test/integration/smoke/test_vpc_ipv6.py:303]: if `update()` throws an exception ...
```suggestion
        self.cleanup.append(network_offering)
        network_offering.update(self.apiclient, state='Enabled')
```

**Review by harikrishna-patnala — APPROVED** [review]
LGTM

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Review by rohityadavcloud — APPROVED** [review]

**Comment by rohityadavcloud** [comment]
@blueorangutan package

File: test/integration/smoke/test_vpc_ipv6.py [hunk 18]
```diff
@@ -0,0 +1,901 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
```
Comment (rohityadavcloud) [inline test/integration/smoke/test_vpc_ipv6.py:4]: @shwstppr does it make sense to move the test_network_ipv6 to smoke/ ? How much time do they take? (asking as the test_vpc_ipv6 is in smoke but the other one is in component/ )

Comment (shwstppr) [reply test/integration/smoke/test_vpc_ipv6.py:4]: @rohityadavcloud There are four test files added,
`smoke/test_network_ipv6.py` - contains 1 test-case
`smoke/test_vpc_ipv6.py` - contains 1 test-case
`component/test_network_ipv6.py` - contains 4 test-cases for upgrade network, redundant VR, etc
`component/test_vpc_ipv6.py` - contains 4 test-cases for upgrade VPC, redundant VR, etc

Each component test file may take over 30mins to execute

Comment (rohityadavcloud) [reply test/integration/smoke/test_vpc_ipv6.py:4]: Thanks @shwstppr

**Comment by rohityadavcloud** [comment]
LGTM, tests have passed, I think xenserver would be more or less same as Travis/build, and tests on kvm, vmware has passed. This only adds tests to the branch.

**Pull request merged**